src/Agent.Worker/Build/BuildJobExtension.cs

```diff
@@ -166,6 +166,39 @@ public override void InitializeJobExtension(IExecutionContext executionContext,
             string pipelineWorkspaceDirectory = Path.Combine(_workDirectory, trackingConfig.BuildDirectory);
 
             UpdateCheckoutTasksAndVariables(executionContext, steps, pipelineWorkspaceDirectory);
+            
+            string selfRepoPath = null;
+            // For saving backward compatibility with the behavior of the Build.RepoLocalPath that was before this PR https://github.com/microsoft/azure-pipelines-agent/pull/3237
+            // We need to change how we set the default value of this variable
+            // We need to allow the setting of paths from RepositoryTrackingInfo for checkout tasks where path input was provided by the user
+            // and this input is not point to the default location for this repository
+            // This is the only case where the value of Build.RepoLocalPath variable is not pointing to the root of sources directory /s.
+            // The new logic is not affecting single checkout jobs and jobs with multiple checkouts and default paths for Self repository
+            if (RepositoryUtil.HasMultipleCheckouts(executionContext.JobSettings))
+            {
+                // get checkout task for self repo 
+                var selfCheckoutTask = steps.Where(x => x.IsCheckoutTask())
+                    .Select(x => x as TaskStep)
+                    .Where(task => task.Inputs.TryGetValue(PipelineConstants.CheckoutTaskInputs.Repository, out string repositoryAlias)
+                                    && RepositoryUtil.IsPrimaryRepositoryName(repositoryAlias)).First();
+                // Check if the task has path input with custom path, if so we need to set as a value of selfRepoPath the value of SourcesDirectory from RepositoryTrackingInfo
+                string path;
+                string selfRepoName = RepositoryUtil.GetCloneDirectory(repoInfo.PrimaryRepository.Properties.Get<string>(Pipelines.RepositoryPropertyNames.Name));
+                if (selfCheckoutTask.Inputs.TryGetValue(PipelineConstants.CheckoutTaskInputs.Path, out path)
+                     && !string.Equals(Path.GetFullPath(Path.Combine(trackingConfig.BuildDirectory, path)),
```

Comment:
Do we need this check? Shouldn't it be null when not specified?

Reply:
Yes, this check is required for handling the case when input contains the path to the default location of the repository which is _/s/RepoName_ since without this condition RepoClean variable will point to the _/s/RepoName_ directory while the previous agent versions will point to the _/s_ directory in this case.